pkg/controller/autoscaling/elasticsearch/resources/resources.go

```diff
@@ -0,0 +1,283 @@
+type ClusterResources []NodeSetsResources
+
+// Match returns true if the resources assigned to a container in a NodeSet matches the one specified in the NodeSetsResources.
+// It returns false if the container is not found in the NodeSet.
+func (ntr NodeSetsResources) Match(containerName string, nodeSet v1.NodeSet) (bool, error) {
+	for _, nodeSetNodeCount := range ntr.NodeSetNodeCount {
+		if nodeSetNodeCount.Name != nodeSet.Name {
+			continue
+		}
+		if nodeSetNodeCount.NodeCount != nodeSet.Count {
+			// The number of nodes in the NodeSetsResources and in the nodeSet is not equal.
+			return false, nil
+		}
+
+		// Compare volume request
+		switch len(nodeSet.VolumeClaimTemplates) {
+		case 0:
+			// If there is no VolumeClaimTemplate in the NodeSet then there should be no storage request in the NodeSetsResources.
+			if ntr.HasRequest(corev1.ResourceStorage) {
+				return false, nil
+			}
+		case 1:
+			volumeClaim := nodeSet.VolumeClaimTemplates[0]
+			if !ResourceEqual(corev1.ResourceStorage, ntr.NodeResources.Requests, volumeClaim.Spec.Resources.Requests) {
+				return false, nil
+			}
+		default:
+			return false, fmt.Errorf("only 1 volume claim template is allowed when autoscaling is enabled, got %d in nodeSet %s", len(nodeSet.VolumeClaimTemplates), nodeSet.Name)
+		}
+
+		// Compare CPU and Memory requests
+		container := getContainer(containerName, nodeSet.PodTemplate.Spec.Containers)
+		if container == nil {
+			return false, nil
+		}
+		return ResourceEqual(corev1.ResourceMemory, ntr.NodeResources.Requests, container.Resources.Requests) &&
+			ResourceEqual(corev1.ResourceCPU, ntr.NodeResources.Requests, container.Resources.Requests), nil
+	}
+	return false, nil
+}
+
+func ResourceEqual(resourceName corev1.ResourceName, expected, current corev1.ResourceList) bool {
+	if len(expected) == 0 {
+		// No value expected, return true
+		return true
+	}
+	expectedValue, hasExpectedValue := expected[resourceName]
+	if !hasExpectedValue {
+		// Expected values does not contain the resource
+		return true
+	}
+	if len(current) == 0 {
+		// Value is expected but current is nil or empty
+		return false
+	}
+	currentValue, hasCurrentValue := current[resourceName]
+	if !hasCurrentValue {
+		// Current values does not contain the resource
+		return false
+	}
+	return expectedValue.Equal(currentValue)
+}
+
+func getContainer(name string, containers []corev1.Container) *corev1.Container {
+	for i := range containers {
+		container := containers[i]
+		if container.Name == name {
+			// Remove the container
+			return &container
+		}
+	}
+	return nil
+}
+
+// NodeSetNodeCount models the number of nodes expected in a given NodeSet.
+type NodeSetNodeCount struct {
+	// NodeSet name.
+	Name string `json:"name"`
+	// NodeCount is the number of nodes, as computed by the autoscaler, expected in this NodeSet.
+	NodeCount int32 `json:"nodeCount"`
+}
+type NodeSetNodeCountList []NodeSetNodeCount
```

Comment:
Would it makes sense to have this as a map[string]int32 to begin with instead of having a custom struct and converting to maps on the fly?

Reply:
The aim of the `NodeSetNodeCount` struct is to make it easier to serialize the status with the pattern `{ name: "...", .... }`, for example:

```
	"nodeSets": [
        {
		"name": "nodeset-di-a",
		"nodeCount": 4
	},
        {
		"name": "nodeset-di-b",
		"nodeCount": 3
	}]
```

I wanted to be consistent with other items in the specs and explicitly have a `name` field in the status.